File: java/client/src/test/java/org/signal/libsignal/net/AuthUsernamesServiceTest.kt
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net

import kotlinx.coroutines.test.runTest
import org.signal.libsignal.internal.NativeTestingNice
import org.signal.libsignal.internal.ReserveUsernameHashOut
import org.signal.libsignal.internal.SetUsernameLinkOut
import org.signal.libsignal.net.assertNonSuccess
import java.util.UUID
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertIs

class AuthUsernamesServiceTest {
  @Test
  fun testReserveUsernameHashName() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_ReserveUsernameHashTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthUsernamesService,
        invoke = { chat, req ->
          chat.reserveUsernameHash(
            usernameHashes = req.usernames,
          )
        },
        check = { expected, actual ->
          when (expected) {
            is ReserveUsernameHashOut.Success ->
              assertContentEquals(
                expected._0,
                assertIs<RequestResult.Success<UsernameHash>>(actual).result,
              )
            ReserveUsernameHashOut.UsernameNotAvailable ->
              actual.assertNonSuccess<_, _, UsernameNotAvailableException>()
          }
        },
      )
    }

  @Test
  fun testSetUsernameLink() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_SetUsernameLinkTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthUsernamesService,
        invoke = { chat, req ->
          chat.setUsernameLink(
            usernameCiphertext = req.usernameCiphertext,
            keepLinkHandle = req.keepLinkHandle,
          )
        },
        check = { expected, actual ->
          when (expected) {
            is SetUsernameLinkOut.Success ->
              assertEquals(
                expected._0,
                assertIs<RequestResult.Success<UUID>>(actual).result,
              )
            SetUsernameLinkOut.UsernameNotSet -> actual.assertNonSuccess<_, _, UsernameNotSetException>()
          }
        },
      )
    }

  @Test
  fun testDeleteUsernameHash() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_DeleteUsernameHashTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthUsernamesService,
        invoke = { chat, _ ->
          chat.deleteUsernameHash()
        },
        check = { _, actual ->
          assertIs<RequestResult.Success<Unit>>(actual)
        },
      )
    }

  @Test
  fun testDeleteUsernameLink() =
    runTest {
      GrpcTestCase.runTests(
        NativeTestingNice.TESTING_DeleteUsernameLinkTests(),
        AuthenticatedChatConnection::fakeConnect,
        ::AuthUsernamesService,
        invoke = { chat, _ ->
          chat.deleteUsernameLink()
        },
        check = { _, actual ->
          assertIs<RequestResult.Success<Unit>>(actual)
        },
      )
    }
}

File: node/ts/test/chat/AuthUsernamesServiceTest.ts
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { config, expect, use } from 'chai';
import chaiAsPromised from 'chai-as-promised';

import * as Native from '../../Native.js';
import * as NativeNice from '../../NativeNice.js';
import * as util from '../util.js';
import { AuthUsernamesService } from '../../net.js';
import { connectAuth, defineTestGrpcCases } from './ServiceTestUtils.js';
import { ErrorCode, LibSignalErrorBase } from '../../Errors.js';

use(chaiAsPromised);

util.initLogger();
config.truncateThreshold = 0;

describe('AuthUsernamesService', () => {
  describe('reserveUsernameHash', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_ReserveUsernameHashTests(),
      connectAuth<AuthUsernamesService>,
      async (
        chat: AuthUsernamesService,
        { usernames }: NativeNice.ReserveUsernameHashArgs,
        resp: NativeNice.ReserveUsernameHashOut
      ) => {
        const out = chat.reserveUsernameHash({
          usernameHashes: usernames,
        });
        if (resp === 'usernameNotAvailable') {
          await expect(out)
            .to.eventually.be.rejectedWith(LibSignalErrorBase)
            .and.deep.include({
              code: ErrorCode.UsernameNotAvailable,
            });
        } else {
          expect(await out).to.deep.equal(resp.success);
        }
      }
    );
  });

  describe('confirmUsername', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_ConfirmUsernameTests(),
      connectAuth<AuthUsernamesService>,
      async (
        chat: AuthUsernamesService,
        { username, usernameCiphertext }: NativeNice.ConfirmUsernameArgs,
        resp: NativeNice.ConfirmUsernameOut
      ) => {
        Native.TESTING_EnableDeterministicRngForTesting();
        const out = chat.confirmUsername({
          username,
          usernameCiphertext,
          rng: { __deterministicRngSeedForTesting: 0 },
        });
        if (resp === 'reservationNotFound') {
          await expect(out)
            .to.eventually.be.rejectedWith(LibSignalErrorBase)
            .and.deep.include({
              code: ErrorCode.UsernameReservationNotFound,
            });
        } else if (resp === 'usernameNotAvailable') {
          await expect(out)
            .to.eventually.be.rejectedWith(LibSignalErrorBase)
            .and.deep.include({
              code: ErrorCode.UsernameNotAvailable,
            });
        } else {
          expect(await out).to.deep.equal(resp.success);
        }
      }
    );
  });

  describe('setUsernameLink', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_SetUsernameLinkTests(),
      connectAuth<AuthUsernamesService>,
      async (
        chat: AuthUsernamesService,
        { usernameCiphertext, keepLinkHandle }: NativeNice.SetUsernameLinkArgs,
        resp: NativeNice.SetUsernameLinkOut
      ) => {
        const out = chat.setUsernameLink({
          usernameCiphertext,
          keepLinkHandle,
        });
        if (resp === 'usernameNotSet') {
          await expect(out)
            .to.eventually.be.rejectedWith(LibSignalErrorBase)
            .and.deep.include({
              code: ErrorCode.UsernameNotSet,
            });
        } else {
          expect(await out).to.deep.equal(resp.success);
        }
      }
    );
  });

  describe('deleteUsernameHash', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_DeleteUsernameHashTests(),
      connectAuth<AuthUsernamesService>,
      async (chat: AuthUsernamesService) => {
        await chat.deleteUsernameHash();
      }
    );
  });

  describe('deleteUsernameLink', () => {
    defineTestGrpcCases(
      NativeNice.TESTING_DeleteUsernameLinkTests(),
      connectAuth<AuthUsernamesService>,
      async (chat: AuthUsernamesService) => {
        await chat.deleteUsernameLink();
      }
    );
  });
});
